nixos/modules/programs/direnv.nix

```diff
@@ -0,0 +1,44 @@
+{ lib, config, pkgs, ... }: {
+
+  options.programs.direnv.enable = lib.mkOption {
+    type = lib.types.bool;
+    default = false;
+    example = true;
+    description = ''
+      Whether to enable direnv integration. Takes care of both installation and
+      setting up the sourcing of the shell. Additionally enables nix-direnv
+      integration. Note that you need to logout and login for this change to apply.
+    '';
+  };
+
+  config = lib.mkIf config.programs.direnv.enable {
+
+    environment.systemPackages = with pkgs; [
+      direnv
+      nix-direnv
+    ];
+
+    environment.pathsToLink = [
+      "/share/direnv"
+    ];
+
+    environment.sessionVariables.DIRENV_CONFIG = "/run/current-system/sw/share/direnv";
+
+    programs.bash.interactiveShellInit = ''
+      eval "$(direnv hook bash)"
+    '';
+
+    programs.zsh.interactiveShellInit = ''
+      eval "$(direnv hook zsh)"
+    '';
+
+    programs.fish.interactiveShellInit = ''
+      direnv hook fish | source
+    '';
+
+    # nix options for derivations to persist garbage collection
+    nix.settings.keep-outputs = true;
```

Comment:
I prefer to use lib.mkDefault instead.

Reply:
Thanks for the detailed explanation. I did not know that this setting is highly recommended for direnv.

Reply:
I guess it would only break the GC-rooting functionality of nix-direnv though, not the caching mechanism overall. Relates to https://github.com/NixOS/nixpkgs/pull/192667#issuecomment-1262652503 as well